lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -121,23 +121,22 @@ class window.Turbolinks
     for k,v of options.headers
       xhr.setRequestHeader k, v
 
-    xhr.onload = ->
-      if xhr.status >= 500
-        Turbolinks.fullPageNavigate(url)
-      else
-        Turbolinks.loadPage(url, xhr, options)
-      xhr = null
-
-    xhr.onerror = ->
-      # Workaround for sinon xhr.abort()
-      if xhr.statusText == "abort"
+    new Promise (resolve) => 
+      xhr.onload = ->
+        if xhr.status >= 500
+          Turbolinks.fullPageNavigate(url)
+        else
+          resolve(Turbolinks.loadPage(url, xhr, options))
         xhr = null
-        return
-      Turbolinks.fullPageNavigate(url)
 
-    xhr.send()
+      xhr.onerror = ->
+        # Workaround for sinon xhr.abort()
+        if xhr.statusText == "abort"
+          xhr = null
+          return
```

Comment:
I had skipped the flows that ended with a full page refresh since your js context will die anyway, but maybe we want to reject on those as well just for clarity?